source/manuals/programming-languages/nodejs/index.html.md.erb

```diff
@@ -82,24 +116,339 @@ Keeping tests in the same directory makes them easier to find and more obvious w
+errors and is often more performant. Write functions as expressions that return
+values of a single type.
+
+Side effects in a function can have bad consequences, especially within `map` or
+`reduce`. `this` is better avoided, as it can refer to different objects
+depending on the context in which a function is executed and lead to unexpected
+side-effects.
+
+Remain aware that JavaScript is not just a functional language and you will
+most probably have to mix functional and object-oriented concepts, which can
+be tricky to get right.
+
+
+More guidance:
+
+- [An Introduction to Functional JavaScript](https://www.sitepoint.com/introduction-functional-javascript/)
+- [JavaScript Allongé](https://leanpub.com/javascriptallongesix/read)
+
+
+## Errors
+
+> Make sure you handle all errors
+
+Envisage all error scenarios, in particular in asynchronous callback functions
+or Promises, and have a fallback for any situation. Consider programmer errors
+(bugs) as well as operational errors (arising from external circumstances,
+like a missing file). Bugs that are caught by exceptions should be logged and
+the execution stopped, and the supervisor will restart the process. Use the
+built-in `Error` object instead of custom types. It makes logging easier.
+
+Your application shouldn't trust any input, for example from a file or an
+API. So the application should handle all operational errors and recover from
+them.
+
+## Node.js's HTTP server
+
+> Offload Node's server as much as possible
+
+Don't expose Node's HTTP server to the public. Use a reverse proxy to serve
+static assets, and cache content as much as possible. Implement adequate
+supervising: [pm2](http://pm2.keymetrics.io/) is recommended for Node.js-specific
+servers.
+
+## Transpiling
+
+> Stick to JavaScript
+
+Avoid anything that compiles to JavaScript (except for static type checking, see
+below). Examples to avoid include CoffeeScript, PureScript and [many others](https://github.com/jashkenas/coffeescript/wiki/List-of-languages-that-compile-to-JS).
+
+If you would like to use static typing consider using JavaScript extensions like
+[TypeScript](http://www.typescriptlang.org/), or [Flow](https://flow.org/). This
+has the advantage of making it easier to prevent runtime errors, by adding type
+information to variables or parameters and transpiling the source to JavaScript,
+reporting errors when wrong types are used.
+
+However be aware that there are disadvantages too: augmented source code will no
+longer be executable uncompiled. You may also need some extra type definitions if you
+want to use external libraries.
+
+Generally, anything that departs from the standard JavaScript syntax in a way
+that a Node developer would have trouble reading your code is advised against.
+
+## Frameworks
+
+> Use Express for web applications, avoid lesser-known frameworks
+
+Express is very common for writing web applications that talk to back-end
```

Comment:
We shouldn't necessarily explain each choice in detail. I think its sufficient to say.
```
## Frameworks and Modules
We standardise on common frameworks and modules across GDS to maximise share-ability of code between projects and familiarity for developers across the GDS community.

### Server Framework
We use express.js 4.x as our server framework.
```

Reply:
Agreed. I'll change that.